cmd/isp/src/cmd.rs

```diff
@@ -0,0 +1,218 @@
+        port,
+        CommandTag::KeyProvision,
+        ResponseCode::Generic,
+        args,
+        DataPhase::NoData,
+    )?;
+
+    Ok(())
+}
+
+pub fn do_enroll(port: &mut dyn serialport::SerialPort) -> Result<()> {
+    let args = vec![
+        // Arg =  Enroll
+        KeyProvisionCmds::Enroll as u32,
+    ];
+
+    do_command(
+        port,
+        CommandTag::KeyProvision,
+        ResponseCode::Generic,
+        args,
+        DataPhase::NoData,
+    )?;
+
+    Ok(())
+}
+
+pub fn do_generate_uds(port: &mut dyn serialport::SerialPort) -> Result<()> {
+    let args = vec![
+        // Arg 0 =  SetIntrinsicKey
+        KeyProvisionCmds::SetIntrinsicKey as u32,
+        // Arg 1 = UDS
+        KeyType::UDS as u32,
+        // Arg 2 = size
+        32_u32,
+    ];
+
+    do_command(
+        port,
+        CommandTag::KeyProvision,
+        ResponseCode::Generic,
+        args,
+        DataPhase::NoData,
+    )?;
+
+    Ok(())
+}
+
+pub fn do_isp_write_keystore(
+    port: &mut dyn serialport::SerialPort,
+    data: Vec<u8>,
+) -> Result<()> {
+    let args = vec![KeyProvisionCmds::WriteKeyStore as u32];
+
+    do_command(
+        port,
+        CommandTag::KeyProvision,
+        ResponseCode::KeyProvision,
+        args,
+        DataPhase::Send { code: ResponseCode::Generic, data: &data },
+    )?;
+
+    Ok(())
+}
+
+pub fn do_isp_read_memory(
+    port: &mut dyn serialport::SerialPort,
+    address: u32,
+    cnt: u32,
+) -> Result<Vec<u8>> {
+    let args = vec![
+        // Arg0 = address
+        address, // Arg1 = length
+        cnt,     // Arg2 = memory type
+        0x0,
+    ];
+
+    let f = do_command(
+        port,
+        CommandTag::ReadMemory,
+        ResponseCode::ReadMemory,
+        args,
+        DataPhase::Recv { code: ResponseCode::Generic, cnt },
+    )?;
+
+    Ok(f.unwrap())
+}
+
+pub fn do_isp_write_memory(
+    port: &mut dyn serialport::SerialPort,
+    address: u32,
+    data: Vec<u8>,
+) -> Result<()> {
+    let args = vec![
+        // arg 0 = address
+        address,
+        // arg 1 = len
+        data.len() as u32,
+        // arg 2 = memory type
+        0x0_u32,
+    ];
+
+    do_command(
+        port,
+        CommandTag::WriteMemory,
+        ResponseCode::Generic,
+        args,
+        DataPhase::Send { code: ResponseCode::Generic, data: &data },
+    )?;
+
+    Ok(())
+}
+
+pub fn do_isp_flash_erase_all(
+    port: &mut dyn serialport::SerialPort,
+) -> Result<()> {
+    let args = vec![
+        // Erase internal flash
+        0x0_u32,
+    ];
+
+    do_command(
+        port,
+        CommandTag::FlashEraseAll,
+        ResponseCode::Generic,
+        args,
+        DataPhase::NoData,
+    )?;
+
+    Ok(())
+}
+
+pub fn do_isp_get_property(
+    port: &mut dyn serialport::SerialPort,
+    prop: BootloaderProperty,
+) -> Result<Vec<u32>> {
+    let args = vec![
+        // Arg 0 = property
+        prop as u32,
+    ];
+
+    send_command(port, CommandTag::GetProperty, args)?;
```

Comment:
If you split `do_command` into multiple arguments like I suggested above, maybe `do_command_nodata` could return the result of `read_response`; then it could be used here and in `do_isp_last_error`